### PR TITLE
Core/Movement: enable item use and spell cast when dismounting

### LINKED ISSUE #6236: Dismount - You can't do this while moving
When mounted if you right click an object or cast a spell it used to dismount fully before attempting the action so the action was successful. Now it seems it tries to do it as you dismount so you get a client error (you can't do this while moving).

2012-04-15 17:41:37 +0200 (39b3589dcb54+) (Win64, Release)
TDB_full_335.11.46_2012_02_13 (+updates)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64867-dismount-you-can-t-do-this-while-moving?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ghost**
**Changes proposed**:

Remove error message and allow spell cast or item use from mounted state on ground, especially when sitting on a flying mount. Current behavior: you get the error message "You can't do that while moving" when trying to cast spells or pick up items while sitting on a flying mount after having landed.

By joschiwald, closes #6236
- allow spell cast or item use from sitting on landed flying mount
- avoid error message from dismount movement check
- works more than 90% better than current behaviour

**Target branch(es)**: 335/6x

**Issues addressed**: Fixes #6236 [**Dismount - You can't do this while moving**](https://github.com/TrinityCore/TrinityCore/issues/6236)

**Tests performed**: (Does it build, tested in-game, etc)
- compiles without errors
- no current known errors in-game

**Known issues and TODO list**:
- [X] extra testing by multiple online users to discover unknown issues
- [X] extended feedback from the TC developers

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -17686,16 +17686,9 @@ bool Unit::SetDisableGravity(bool disable, bool /*packetOnly = false*/)
...
         RemoveUnitMovementFlag(MOVEMENTFLAG_DISABLE_GRAVITY);
-        if (!HasUnitMovementFlag(MOVEMENTFLAG_CAN_FLY))
-        {
-            m_movementInfo.SetFallTime(0);
-            AddUnitMovementFlag(MOVEMENTFLAG_FALLING);
-        }
-    }
 
-    return true;
+           return true;
```
Comment (Nawuko) [inline src/server/game/Entities/Unit/Unit.cpp]: Wrong spacing

Comment (tkrokli) [inline src/server/game/Entities/Unit/Unit.cpp]: Thanks, I was unaware of that. I thought I had followed joschiwald's diff precisely.

File: src/server/game/Movement/MotionMaster.cpp [hunk 2]
```diff
@@ -479,11 +479,12 @@ void MotionMaster::MoveFall(uint32 id /*=0*/)
     if (std::fabs(_owner->GetPositionZ() - tz) < 0.1f)
         return;
 
+    _owner->AddUnitMovementFlag(MOVEMENTFLAG_FALLING);
```
Comment (Shauren) [inline src/server/game/Movement/MotionMaster.cpp]: All changes made to this file are wrong. Creatures should never receive MOVEMENTFLAG_FALLING

**Comment by tkrokli** [comment]
By the way, I removed some extra blank lines after each `RemoveUnitMovementFlag()` in the same section of `Unit.cpp`. Do you know if that will have any negative consequences, that those blank lines should be kept in the file like they were before my last edit?

File: src/server/game/Movement/MotionMaster.cpp (hunk 2 above)
Comment (tkrokli) [inline src/server/game/Movement/MotionMaster.cpp]: Just after my previous comment in the [Conversations] tab, I received an e-mail with a comment from Shauren, added to this point in the file, saying: "All changes made to this file are wrong. Creatures should never receive MOVEMENTFLAG_FALLING". Does that comment still hold true? (Maybe github somehow removed that comment and it should not have been removed? I don't know.)

Comment (Shauren) [inline src/server/game/Movement/MotionMaster.cpp]: I deleted it myself, forget it.

**Comment by Aokromes** [comment]
Any reason to don't merge this?

**Comment by tkrokli** [comment]
Maybe I should set joschiwald as author, it is after all his content. BRB with the update.

**Comment by Nelidon** [comment]
Can be merged now.

**Comment by tkrokli** [comment]
@Nelidon : thanks for the input. I guess several people have been using this for a while, but do you know if it has been tested in a multi-user scenario? :)

**Comment by Nelidon** [comment]
Give me few days to test it with ~25 connected.

**Comment by mik1893** [comment]
this should be merged - anyone against it?

**Comment by mik1893** [comment]
"Qui tacet consentire videtur"

**Comment by Nelidon** [comment]
We have again the same issue.

**Comment by tkrokli** [comment]
Please open a new issue for it and **include all the requested info** in the TC issue template.

**Pull request merged**